### PR TITLE
[android][mail-composer] Rewrite mail composer module from Java to Kotlin

### PR DESCRIPTION
**Pull request by M1ST4KE**
# Why

Modules written in Java are being rewritten to Kotlin

# How

- Generated Kotlin code from java.
- Clean up code.

# Test Plan

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  } catch (e: Exception) {
+    Uri.fromFile(file)
+  }
+
+  fun build() = mailIntent
+
+  fun setComponentName(pkg: String, cls: String): MailIntentBuilder {
+    mailIntent.component = ComponentName(pkg, cls)
+    return this
+  }
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt:37]:
```suggestion
  fun setComponentName(pkg: String, cls: String): MailIntentBuilder = run {
    mailIntent.component = ComponentName(pkg, cls)
  }
```

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  fun putExtraIfKeyExists(key: String, intentName: String): MailIntentBuilder {
+    if (options.containsKey(key)) {
+      if (options.getList(key) != null) {
+        mailIntent.putExtra(intentName, getStringArrayFrom(key))
+      } else {
+        mailIntent.putExtra(intentName, options.getString(key))
+      }
+    }
+    return this
+  }
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt:48]:
```suggestion
  fun putExtraIfKeyExists(key: String, intentName: String): MailIntentBuilder = run {
    if (options.containsKey(key)) {
      if (options.getList(key) != null) {
        mailIntent.putExtra(intentName, getStringArrayFrom(key))
      } else {
        mailIntent.putExtra(intentName, options.getString(key))
      }
    }
  }
```

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    if (options.containsKey(key)) {
+      val body = if (isBodyHtml) {
+        Html.fromHtml(options.getString(key))
+      } else {
+        options.getString(key)
+      }
+      mailIntent.putExtra(intentName, body)
+    }
+    return this
+  }
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt:60]:
```suggestion
  fun putExtraIfKeyExists(key: String, intentName: String, isBodyHtml: Boolean): MailIntentBuilder = run {
    if (options.containsKey(key)) {
      val body = if (isBodyHtml) {
        Html.fromHtml(options.getString(key))
      } else {
        options.getString(key)
      }
      mailIntent.putExtra(intentName, body)
    }
  }
```

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          val attachmentFile = File(path)
+          contentUriFromFile(attachmentFile, application)
+        }.toCollection(ArrayList())
+        mailIntent.putParcelableArrayListExtra(intentName, attachments)
+      }
+    } catch (error: IllegalArgumentException) {
+      Log.e("ExpoMailComposer", "Illegal argument:", error)
+    }
+    return this
+  }
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt:82]:
```suggestion
  fun putParcelableArrayListExtraIfKeyExists(
    key: String,
    intentName: String,
    application: Application,
  ): MailIntentBuilder = run {
    try {
      if (options.containsKey(key)) {
        val requestedAttachments = getStringArrayFrom(key)
        val attachments = requestedAttachments.map { requestedAttachment ->
          val path = Uri.parse(requestedAttachment).path
          requireNotNull(path, { "Path to attachment can not be null" })
          val attachmentFile = File(path)
          contentUriFromFile(attachmentFile, application)
        }.toCollection(ArrayList())
        mailIntent.putParcelableArrayListExtra(intentName, attachments)
      }
    } catch (error: IllegalArgumentException) {
      Log.e("ExpoMailComposer", "Illegal argument:", error)
    }
  }
```

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import expo.modules.core.arguments.ReadableArguments
+import java.io.File
+
+class MailIntentBuilder(
+  private val options: ReadableArguments
+) {
+  private val mailIntent = Intent(Intent.ACTION_SEND_MULTIPLE)
+
+  private fun getStringArrayFrom(key: String): Array<String?> {
+    return (options.getList(key) as List<String?>).toTypedArray()
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt:19]: It is an unchecked cast, isn't it? Suppress the warning.

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @ExpoMethod
+  fun isAvailableAsync(promise: Promise) {
+    promise.resolve(true)
+  }
+
+  @ExpoMethod
+  fun composeAsync(options: ReadableArguments, promise: Promise) {
+    val intent = Intent(Intent.ACTION_SENDTO)
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt:39]:
```suggestion
    val intent = Intent(Intent.ACTION_SENDTO).apply { data =  Uri.parse("mailto:") }
```

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    val application = activityProvider.currentActivity.application
+    intent.data = Uri.parse("mailto:")
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt:41]:
```suggestion
```

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+          Intent.EXTRA_STREAM,
+          application
+        )
+      LabeledIntent(
+        mailIntentBuilder.build(),
+        info.activityInfo.packageName,
+        info.loadLabel(context.packageManager),
+        info.icon
+      )
+    }.toMutableList()
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt:63]:
```suggestion
    }.toList()
```
❓

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    pendingPromise = promise
+    val chooser = Intent.createChooser(
+      mailIntents.removeAt(mailIntents.size - 1),
+      null
+    ).apply {
+      putExtra(Intent.EXTRA_INITIAL_INTENTS, mailIntents.toTypedArray())
+      addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+      addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
+    }
```
Comment (lukmccall) [inline packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt:72]:
```suggestion
    val chooser = Intent.createChooser(
      mailIntents.removeAt(mailIntents.size - 1),
      null
    ).apply {
      putExtra(Intent.EXTRA_INITIAL_INTENTS, mailIntents.toTypedArray())
      addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
    }
    pendingPromise = promise
```

**Review by lukmccall — CHANGES_REQUESTED** [review]

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt (hunk 8 above)
Comment (M1ST4KE) [reply packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailComposerModule.kt:62]: Look at line 66 in this file.

File: packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt (hunk 1 above)
Comment (M1ST4KE) [reply packages/expo-mail-composer/android/src/main/java/expo/modules/mailcomposer/MailIntentBuilder.kt]: I suppose you meant `apply` instead of `run` ;)

**Review by bbarthec — APPROVED** [review]
As for me it looks fine 👍

**Review by lukmccall — APPROVED** [review]
LGMT ✅  but please fix the spotless error

**Pull request merged**